libcloudforensics/aws.py

```diff
@@ -0,0 +1,694 @@
+      volume_name = self._GenerateVolumeName(
+          snapshot, volume_name_prefix=volume_name_prefix)
+
+    if not REGEX_TAG_VALUE.match(volume_name):
+      raise ValueError('Volume name {0:s} does not comply with '
+                       '{1:s}'.format(volume_name, REGEX_TAG_VALUE.pattern))
+
+    client = self.ClientApi(EC2_SERVICE)
+    try:
+      volume = client.create_volume(
+          AvailabilityZone=snapshot.availability_zone,
+          SnapshotId=snapshot.snapshot_id,
+          TagSpecifications=[GetTagForResourceType('volume', volume_name)]
+      )
+      volume_id = volume['VolumeId']
+      zone = volume['AvailabilityZone']
+      # Wait for volume creation completion
+      client.get_waiter('volume_available').wait(VolumeIds=[volume_id])
+    except (client.exceptions.ClientError,
+            botocore.exceptions.WaiterError) as exception:
+      raise RuntimeError('Could not create volume {0:s} from snapshot '
+                         '{1:s}: {2:s}'.format(volume_name, snapshot.name,
+                                               str(exception)))
+
+    return AWSVolume(volume_id, self, self.default_region, zone,
+                     name=volume_name)
+
+  def _GenerateVolumeName(self, snapshot, volume_name_prefix=None):
+    """Generate a new volume name given a volume's snapshot.
+
+    Args:
+      snapshot (AWSSnapshot): A volume's Snapshot.
+      volume_name_prefix (str): Optional. Prefix for the volume name.
+
+    Returns:
+      str: A name for the volume.
+
+    Raises:
+      ValueError: if the volume name does not comply with the RegEx.
+    """
+
+    # Max length of tag values in AWS is 255 characters
+    # UserId is expected to be set if the call to the EC2 API is successful.
+    # https://boto3.amazonaws.com/v1/documentation/api/1.9.42/reference/services/sts.html#STS.Client.get_caller_identity for more details. # pylint: disable=line-too-long
+    user_id = self.ClientApi(ACCOUNT_SERVICE).get_caller_identity()['UserId']
+    volume_id = user_id + snapshot.volume.volume_id
+    volume_id_crc32 = '{0:08x}'.format(
+        binascii.crc32(volume_id.encode()) & 0xffffffff)
+    truncate_at = 255 - len(volume_id_crc32) - len('-copy') - 1
+    if volume_name_prefix:
+      volume_name_prefix += '-'
+      if len(volume_name_prefix) > truncate_at:
+        # The volume name prefix is too long
+        volume_name_prefix = volume_name_prefix[:truncate_at]
+      truncate_at -= len(volume_name_prefix)
+      volume_name = '{0:s}{1:s}-{2:s}-copy'.format(
+          volume_name_prefix, snapshot.name[:truncate_at], volume_id_crc32)
+    else:
+      volume_name = '{0:s}-{1:s}-copy'.format(
+          snapshot.name[:truncate_at], volume_id_crc32)
+
+    return volume_name
+
+  def GetInstanceById(self, instance_id, region=None):
+    """Get an instance from an AWS account by its ID.
+
+    Args:
+      instance_id (str): The instance id.
+      region (str): Optional. The region to look the instance in.
```

Comment:
What happens when the caller doesn't specify a region? Will it look in all regions? Can more than one instance be returned in that case? (or do instances have a unique ID per region?). Same question for other functions

Reply:
If not specified it will use `self.default_region` which is derived from `self.default_availability_zone`, that is given by the user when instantiating `AWSAccount`. It will therefore always look in one region only, so it'll return only one instance if the ID exists in that region. This is a limitation of the `boto3` API, as instances of `boto3.session.Session().client()` must specify exactly one region.

Reply:
Ok I see. Could you please add this to the docstring?